lib/SIL/SILConstants.cpp

```diff
@@ -0,0 +1,395 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/SIL/SILConstants.h"
+#include "swift/AST/DiagnosticsSIL.h"
+#include "swift/Demangling/Demangle.h"
+#include "swift/SIL/SILBuilder.h"
+#include "llvm/Support/TrailingObjects.h"
+using namespace swift;
+
+template <typename... T, typename... U>
+static InFlightDiagnostic diagnose(ASTContext &Context, SourceLoc loc,
+                                   Diag<T...> diag, U &&... args) {
+  return Context.Diags.diagnose(loc, diag, std::forward<U>(args)...);
+}
+
+//===----------------------------------------------------------------------===//
+// SymbolicValue implementation
+//===----------------------------------------------------------------------===//
+
+void SymbolicValue::print(llvm::raw_ostream &os, unsigned indent) const {
+  os.indent(indent);
+  switch (representationKind) {
+  case RK_Unknown: {
+    os << "unknown(" << (int)getUnknownReason() << "): ";
+    getUnknownNode()->dump();
+    return;
+  }
+  case RK_Metatype:
+    os << "metatype: ";
+    getMetatypeValue()->print(os);
+    os << "\n";
+    return;
+  case RK_Function: {
+    auto fn = getFunctionValue();
+    os << "fn: " << fn->getName() << ": ";
+    os << Demangle::demangleSymbolAsString(fn->getName());
+    os << "\n";
+    return;
+  }
+  case RK_Integer:
+  case RK_IntegerInline:
+    os << "int: " << getIntegerValue() << "\n";
+    return;
+  case RK_Aggregate: {
+    ArrayRef<SymbolicValue> elements = getAggregateValue();
+    switch (elements.size()) {
+    case 0:
+      os << "agg: 0 elements []\n";
+      return;
+    case 1:
+      os << "agg: 1 elt: ";
+      elements[0].print(os, indent + 2);
+      return;
+    default:
+      os << "agg: " << elements.size() << " elements [\n";
+      for (auto elt : elements)
+        elt.print(os, indent + 2);
+      os.indent(indent) << "]\n";
+      return;
+    }
+  }
+  }
+}
+
+void SymbolicValue::dump() const { print(llvm::errs()); }
+
+/// For constant values, return the classification of this value.  We have
+/// multiple forms for efficiency, but provide a simpler interface to clients.
+SymbolicValue::Kind SymbolicValue::getKind() const {
+  switch (representationKind) {
+  case RK_Unknown:
+    return Unknown;
+  case RK_Metatype:
+    return Metatype;
+  case RK_Function:
+    return Function;
+  case RK_Aggregate:
+    return Aggregate;
+  case RK_Integer:
+  case RK_IntegerInline:
+    return Integer;
+  }
+}
+
+/// Clone this SymbolicValue into the specified allocator and return the new
+/// version.  This only works for valid constants.
+SymbolicValue
+SymbolicValue::cloneInto(llvm::BumpPtrAllocator &allocator) const {
+  auto thisRK = representationKind;
+  switch (thisRK) {
+  case RK_Unknown:
+  case RK_Metatype:
+  case RK_Function:
+  case RK_IntegerInline:
```

Comment:
@marcrasi  Seems like this function will have an assert failure inside `getIntegerValue()` if we call it with any of the first four representation kinds.  I suppose there is no reason why one should clone the first four cases as their values will probably not change at all.  In that case, can we have a default case that fails with an assert and a reasonable error message (like "cloning of this representation kind ... is not supported"? Then we can just remove the first four cases.